Handle migrated celo headers

piersy opened the pull request:
This PR updates the header encoding and decoding for celo blocks to be able to support data migrated from celo-blockchain.

Our post gingerbread headers actually decode into the op-geth header structure without any changes. Since adding fields to the end of a struct does not cause rlp decoding to fail for previously encoded values. Our pre-gingerbread don't decode into the op-geth header structure because the pre-gingerbread headers miss fields from the middle of the struct so we need to handle those explicitly.

The decision on how to decode is made based on introspecting the rlp and looking at the second field which is a hash (length 32 bytes) post-gingerbread, but an address (length 20 bytes) pre-gingerbread.

The decision on how to encode is made based on the value of gasLimit if zero then it's a pre-gingerbread header.

Additionally we needed to allow genesis blocks in a CeL2 migrated chain context (by checking if `Cel2Time` is set and non zero) that lack GasLimit and Difficulty because those fields were considered required by op-geth but not present in early Celo.

Finally some tests that used blocks with zero gas limits needed to be updated since with a zero gas limit the different encoding was now breaking them, and adding a non zero gas limit meant the block hash changed.

piersy committed:
Add custom decoding for celo headers

This allows op geth to decode old celo headers.

piersy committed:
Update header decoding to introspect the rlp data

This presents a more efficient approach to determining if an encoded
header is pre or post gingerbread.

piersy committed:
Add encoding for pre-gingerbread headers

piersy committed:
Move extra celo code to celo_block.go

Also remove unused header fields

piersy committed:
Allow genesis without difficulty and gas limit

In Celo's case genesis blocks did not have difficulty or gas limit.

piersy committed:
Fix TestRPCMarshalBlock

We needed to add a non zero gas limit to ensure the block is encoded in
post gingerbread form.

piersy committed:
Fix TestRPCGetBlockOrHeader

In order for the block headers to serailise to the op-geth structure we
needed to set a non zero gasLimit.

piersy committed:
Fix TestNotify

In order for the header to serailise to the op-geth structure we needed
to set a non zero gasLimit.

piersy committed:
Prevent ToBlock overwriting difficulty & gasLimit

In ethereum's case these settings were senisble defaults but celo did
not start with gasLimit or Difficulty in the genesis block, so we musn't
overwrite them when unset.

palango commented:
> The decision on how to encode is made based on the value of gasLimit if zero then it's a pre-gingerbread header.

I thought we would set that to the correct value, so that the apis return something useful. Or is this done somewhere in the API layer?

> Additionally we needed to allow genesis blocks in a CeL2 context (by checking if Cel2Time is set) that lack GasLimit and Difficulty because those fields were considered required by op-geth but not present in early Celo.

Is this requires so that the genesis hash stays the same, or why can't we just add those values to the celo genesis block?

> Many tests using TestChainConfig failed because it set Cel2Time and that caused genesis specs without gas limit or difficulty set to not have a default applied in Genesis.ToBlock

For this you can use `TestChainConfigNoCel2`, so that you don't need to touch those tests.

palango commented on params/config.go:
```diff
@@ -269,7 +269,6 @@ var (
 		CancunTime:                    nil,
 		PragueTime:                    nil,
 		VerkleTime:                    nil,
-		Cel2Time:                      newUint64(0),
```
Please revert this, we want to enable `Cel2` for as many tests as possible.

If a test fails please opt-out by using `TestChainConfigNoCel2`.

piersy replied:
The problem is that enabling Cel2 now causes many tests to fail:

Many tests do not specify gasLimit or difficulty in the geneses they create, assuming that those fields will be filled in by `Genesis.ToBlock` but if Cel2 is enabled they won't be filled in and then they fail. To fix all those tests we'd end up with a lot of changes scattered throughout the codebase.

It seems to me that the only tests that really need Cel2 enabled are our tests that are testing the Cel2 functionality, the rest can be run without Cel2 enabled.

piersy replied:
Ok I've raised a PR [here](https://github.com/celo-org/op-geth/pull/147) that reverts this change and makes use of `TestChainConfigNoCel2`.  It's not as much changes as I was envisaging but it's still quite a lot of extra scattered changes which will make updating from upstream more difficult. 

If the approach to fixing failed tests is simply to rely on `TestChainConfigNoCel2` then eventually (when all tests are using `TestChainConfigNoCel2`) we'll end up in the same situation as just not setting Cel2Time in the TestChainConfig but with a larger diff.

Also the approach of using `TestChainConfigNoCel2` is not particularly precise because sometimes we are setting it in a utility function and so it's effect is not confined to just the failing test.

So I think we would save ourselves some trouble if we didn't  add `Cel2Time` to the `TestConfig` and instead be aware of our specific changes and ensure that we test around them appropriately.

What do you think?

Just for reference, current changes we have that rely on `Cel2Time` are:

* Not setting gasLimit and difficulty to defaults in `genesis.ToBlock`
* Selecting transaction signers
* Setting exchange rates field in the block context
* Checking that the fee currency is whitelisted during the state transition
* Selecting the fee handler as the recipient of the base fee
* Adding celo precompiles to the set of precompiles acessible by the vm

palango replied:
I still think it's worth having Cel2 flag enabled by default, so that we find potential errors when merging upstream.

I'd go with those changes that you did in the other branch. If we run into lot's of problems in the future during rebases, we can still change this.

piersy replied:
Ok will merge that other branch in here.

piersy replied:
Ok so actually this turned out to be a much larger task than it originally looked. Locally the go test caching failed me showing that tests were passing for https://github.com/celo-org/op-geth/pull/147, but actually they were not.

One of the problems with this is that the code panics which stops execution of the tests in a package, so you can't easily find all the failing tests, you have to fix one and then try again. Also now the diff is 4 times larger than #147 and tests are still failing, its not clear how much more will need to be changed.

So my feeling is that setting TestChainConfigNoCel2 on an individual test basis is not a sustainable approach.

@carterqw2 do you have any thoughts?

palango commented on core/gen_genesis.go:
```diff
@@ -64,22 +64,22 @@ func (g Genesis) MarshalJSON() ([]byte, error) {
+		Difficulty    *math.HexOrDecimal256                      `json:"difficulty"`
+		Mixhash       *common.Hash                               `json:"mixHash"`
+		Coinbase      *common.Address                            `json:"coinbase"`
+		Alloc         map[common.UnprefixedAddress]types.Account `json:"alloc"      gencodec:"required"`
```
Any idea why the type changes here? We should probably upstream that.

piersy replied:
Hmm, no idea I ran `go generate` after removing required on gas limit and difficulty and this was the result.

piersy replied:
But GenesisAccount is an alias for types.Account so I don't think there should be any problem with this.

palango replied:
This is already changes upstream, so is fine.

palango commented on core/types/gen_header_rlp.go:
```diff
@@ -1,85 +0,0 @@
-// Code generated by rlpgen. DO NOT EDIT.
```
Why is this removed?

piersy replied:
Because we define our own EncodeRLP in `celo_block.go` that handles pre and post gingerbread rlp encodings.

palango replied:
Sure, but there it's just `rlp.Encode(w, &encodedHeader)` to encode it. If it's that easy, wthen why did all of this code exist? Will this have an impact on performance?

piersy replied:
Yeah it was performance see here - https://github.com/ethereum-optimism/op-geth/commit/d6f49bf764c48e9415b168e701b6dc26566a675c

So we could probably add encoding like this back in for the 2 types we support. But I'd rather leave that for later since right now performance is not an issue.

piersy replied:
Added https://github.com/celo-org/op-geth/issues/145

piersy commented:
>> The decision on how to encode is made based on the value of gasLimit if zero then it's a pre-gingerbread header.

>I thought we would set that to the correct value, so that the apis return something useful. Or is this done somewhere in the API layer?

So we decided that we wouldn't add this to the encoded representation since it then breaks the assumption that `hash(rlp(header)) == header.Hash()`. In fact we strip out the `IstanbulAggregatedSeal` during migration so that `hash(rlp(header)) == header.Hash()`.

We can still return this at the api layer, but we are not going to put it in the rlp encoded header representation.

>> Additionally we needed to allow genesis blocks in a CeL2 context (by checking if Cel2Time is set) that lack GasLimit and Difficulty because those fields were considered required by op-geth but not present in early Celo.

>Is this requires so that the genesis hash stays the same, or why can't we just add those values to the celo genesis block?

It is precisely for the genesis hash, if those fields are set then the hash changes and then you can't connect to the network due to a genesis hash mismatch.

>> Many tests using TestChainConfig failed because it set Cel2Time and that caused genesis specs without gas limit or difficulty set to not have a default applied in Genesis.ToBlock

>For this you can use `TestChainConfigNoCel2`, so that you don't need to touch those tests.

Ok can have a look at how that works out.

alecps commented on core/types/celo_block.go:
```diff
@@ -0,0 +1,153 @@
+func (h *Header) EncodeRLP(w io.Writer) error {
+	// We check for a pre gingerbread header by looking for (GasLimit == 0)
+	// here. We don't use Difficulty  because CopyHeader can end up setting a
+	// nil Difficulty to a zero difficulty, so testing for nil difficulty is
```
Why does setting a nil difficulty to a zero difficulty cause a problem? We are checking for a zero gaslimit, not a nil gaslimit, so I would think they would be the same right?

piersy replied:
Post gingerbread the difficulty is hardcoded to zero, so if we were to use `difficulty == 0` to select the encoding pre-gingerbread headers copied by `CopyHeader` would be encoded in a post gingerbread form.

piersy committed:
Ensure non nil difficulty in pre-gingerbread blocks

Nil difficulty in headers causes problems because it is assumed by the
go-ethereum codebase that difficulty is non nil.

The specific problem requiring this fix was a NPE in
Downloader.processHeaders where the total difficulty (td) is calculated
by adding individual header difficulty values.

Rather than try to fix this specific case it seems safer to ensure
blocks have a non nil Difficulty which should help future proof since it
upholds the assumptions of the go-ethereum codebase.

piersy committed:
Run tests with Cel2Time set

Rather than remove Cel2Time from TestChainConfig this
PR uses TestChainConfigNoCel2 to selectively remove
Cel2Time from failing tests.

piersy committed:
Make tests versbose to find failing ci tests

piersy committed:
Use TestChainConfigNoCel2 to fix panicking ci tests

piersy committed:
Switch to TestChainConfigNoCel2 in a couple more places

piersy committed:
Revert "Ensure non nil difficulty in pre-gingerbread blocks"

This reverts commit 8eec26f5ddf5687f081bcf5eef9c82868da989f8.

piersy committed:
Add more no cel2 config

piersy committed:
Switch further config to NoCel2

piersy committed:
Use TestChainConfigNoCel2 in miner test

piersy committed:
Hopefully last couple of adjustments

piersy committed:
Maybe the final adjustment

piersy committed:
Is this the last use of no cel 2 ??

piersy committed:
I feel it, this is the one!!

piersy committed:
Another batch

carterqw2 commented on core/genesis.go:
```diff
@@ -475,11 +475,15 @@ func (g *Genesis) ToBlock() *types.Block {
-		head.Difficulty = params.GenesisDifficulty
+	// Disable setting default gas limit and difficulty in cel2 context.
+	// Because the celo genesis lacked gas limit and difficulty.
+	if g.Config.Cel2Time == nil {
```
Can we modify this check to something similar to the one [here](https://github.com/celo-org/op-geth/pull/143/files#diff-1272116434e5549f278a940767af52ed2d1abc91c44fa04ec48d8835a3774036R488) and check if Cel2 was activated in block 0 (which happens only in tests) and set values appropriately?

piersy replied:
Yeah, I think this will work!

piersy replied:
Thank you @carterqw2, that has solved the issue with all the tests!

piersy committed:
Revert "Use TestChainConfigNoCel2 to fix panicking ci tests"

This reverts commit b7993a790465b32ee1348d8177f1622a9d864400.

piersy committed:
Revert "Make tests versbose to find failing ci tests"

This reverts commit 1f8290582568c5322f1edc5d95cb83e555abb0a5.

piersy committed:
Revert "Run tests with Cel2Time set"

This reverts commit 249a0e7a48a5071c54ac4dc39c9bda14f6bc017f.

piersy committed:
Revert "Switch to TestChainConfigNoCel2 in a couple more places"

This reverts commit 1ea65fd2fbbc5bf4099a0eac78ab76e248e96fd8.

piersy committed:
Revert "Another batch"

This reverts commit 1cf71832c8e866d5845415ed1001f7cd3dc1ce79.

piersy committed:
Revert "I feel it, this is the one!!"

This reverts commit 9e98c5c26aad5766b8cc0fc023ade06d7410dde1.

piersy committed:
Revert "Is this the last use of no cel 2 ??"

This reverts commit 2d2bfcc585efc4c69c2ffda607974df45573c1c6.

piersy committed:
Revert "Maybe the final adjustment"

This reverts commit 10dff95f9ac81bfa6b975e41ecd75906ce850536.

piersy committed:
Revert "Hopefully last couple of adjustments"

This reverts commit 4aaa46fd503f7a5589eb69d127f65c5c97aa6fb9.

piersy committed:
Revert "Use TestChainConfigNoCel2 in miner test"

This reverts commit 696d544d169a37a8e7311c7c55736b1d276db360.

piersy committed:
Revert "Switch further config to NoCel2"

This reverts commit 91a67919c0cc6dd01f4680fcf2b370ce777fa05d.

piersy committed:
Revert "Add more no cel2 config"

This reverts commit 18236315b1d3d1df63c4fed5f257dec0cda9243c.

piersy committed:
Revert "Prevent ToBlock overwriting difficulty & gasLimit"

This reverts commit 51d32c2c5472365215767890b534b3a51a9cff23.

piersy committed:
Prevent setting default gasLimit and difficulty for migrated celo chains

piersy committed:
Update Header EncodeRLP comment

piersy committed:
Improve documentation for isPreGingerbreadHeader

piersy committed:
Merge branch 'celo6' into piersy/handle-migrated-blocks-and-headers
